File: coeff/fir_coeff.py
# Generate coefficients for a FIR lowpass or notch filter

from scipy.signal import kaiserord, firwin, freqz

fs = 500;  # sampling freq in Hz
fc = np.array([59, 61]);   # cutoff freq in Hz

ripple_db = 30.0
width = 10 / fs
N, beta = kaiserord(ripple_db, width);
N |= 1  # convert N to odd number
hh = firwin(N, fc/(fs/2), window=('kaiser', beta), pass_zero='bandstop')


# visualize frequency response
w, h = freqz(hh)
plt.semilogx(fs * w / (2*np.pi), 20 * np.log10(np.maximum(abs(h), 1e-5)))
plt.title(f"FIR Filter Frequency Response")
plt.xlabel('Frequency [Hz]')
plt.ylabel('Amplitude [dB]')
plt.grid(which='both', axis='both')
plt.show()

# print out coefficients in C array format
def print_c_array(coeff, name, dtype='float'):
    coeff_str = f"const {dtype} {name} = {{"
    for c in coeff:
        coeff_str += f"{c:.8f},"
    coeff_str += "};"
    print(coeff_str)
print_c_array(hh, 'h')
print(f"filter length = {len(hh)}")

File: coeff/fir_coeff.m
% Generate coefficients for a FIR notch filter
% tonyfu97@uw.edu, Feb 2023
% Credit: https://www.mathworks.com/matlabcentral/answers/1591484-i-want-to-remove-50hz-noise-using-fir-notch-filter-in-ecg-signal

fs = 350;  % sampling freq in Hz
lowpass = false;

if lowpass
    fc = 30;
    n = 31;
    hh = fir1(n, fc./(fs/2), 'low');
else
    fcomb = [55 59 61 64];  % cutoff freq in [pass1 stop1 stop2 pass2] format
    mags = [1 0 1];
    dev = [[0.5 0.1 0.5]];
    [n,Wn,beta,ftype] = kaiserord(fcomb,mags,dev,fs);
    hh = fir1(n,Wn,ftype,kaiser(n+1,beta),'noscale');
end

% visualize frequency response
figure
freqz(hh, 1, 2^20, fs)
set(subplot(2,1,1), 'XLim', [0 100])
set(subplot(2,1,2), 'XLim', [0 100])

% print out coefficients in C array format
disp(['const float h[] = {', sprintf('%0.8f,', hh), '};']);